### DISCUSSION TITLE
SimCLR on custom dataset on a SLURM cluster

### DISCUSSION THREAD

**Discussion by MathiasBaumgartinger** [torchgeo/torchgeo#3516] [Q&A] [answered]
I am trying to re-use the SimCLR implementation of torchgeo on FLAIR2 dataset (yes I am sorry, I think I will finally be able to finalize the Pull-Request soon :sweat_smile: https://github.com/torchgeo/torchgeo/pull/2394). 

Unfortunately, my learning suggest something is not working flawlessly, as the estimated moving average of the `ssl_std` is approaching 0 and training loss seems to stop reducing: 

<img width="1394" height="333" alt="image" src="https://github.com/user-attachments/assets/bc495f43-18c5-4244-afb5-a30adb2cbb56" />

My hyperparams:
```yaml
  temperature: 0.07
  batch_size: 1024
  chip_size: 64
  lr_multiplier: 1. # lr: Learning rate (0.3 x batch_size / 256 is recommended)
  architecture: "resnet50"
  max_steps: 50000
  min_steps: 50000
```

I have tried running the default augmentations and slight variations (mostly less "strong"). The data is normalized (` K.Normalize(0, 255)` and if I plot it it looks fine. I have really tried fiddeling around with the hparams but cannot seem to find a good solution.

I have also set the `sync_batchnorm=True` to avoid leaking cross-batch results. By any chance have you stumbled upon similar issues? How could you solve? 

___

Locally, I have been able to produce better results on smaller subsets of the data or entirely different datasets, the batch-size and data I have used seems to be too small to produce usable downstream weights though:

<img width="1639" height="331" alt="image" src="https://github.com/user-attachments/assets/c08d4f85-338b-4018-95ea-3e8d49d18b66" />

**Comment by adamjstewart** [comment]
@AABNassim ran the SimCLR experiments in our SSL4EO-L paper and also struggled to get good results. At the time, we suspected our batch size was too low to provide sufficient negative samples, but your batch size is quite a bit bigger. 

SSL is very finicky to get right. Our SimCLRTask was also designed to mimic the original SimCLR paper hyperparameters, but newer research may have found better results. 

I would suggest asking the developers of [lightly](https://github.com/lightly-ai/lightly), they were kind enough to comment on some of our PRs to let us know that there may be better defaults.

**Reply by MathiasBaumgartinger** [reply]
Thanks. @AABNassim if you have any hints in which direction I may want to go, please hit me up :-)

**Reply by MathiasBaumgartinger** [reply]
Also @adamjstewart @AABNassim, do you happen to have the training protocols handy? Even just a screenshot of the TensorBoard plots would be incredibly useful for calibration. Thanks!"

**Reply by AABNassim** [reply]
Hi @MathiasBaumgartinger, 

Unfortunately, I don't have the logs anymore. I'm also not sure what's going on with your runs. Some random thoughts:

- Make sure your augmentation pipeline is working as intended. I assume it's already the case since you mentioned visualizing the samples (after the augmentations?). I recall some of the Kornia augmentations were tricky w.r.t to normalization. 
- 64x64 is somewhat small for random resized crop, but I would still expect it to work.
- Negative pairs can be tricky in small/not very diverse RS datasets. You will likely have a lot of negatives that are very similar to your positive pairs. You can try different values for the temperature (e.g., 0.1) and compare the effect on training.
- Learning rate and batch size are definitely hyperparameters to play with if you haven't already.

**Reply by adamjstewart** [reply]
We don't have the logs, but we do have the config files we used: https://github.com/torchgeo/ssl4eo-l/tree/main/landsat/conf

**Reply by MathiasBaumgartinger** [reply]
Do you by any chance recall what the curves _approximately_ looked like? I suppose the `ssl_std` looked somewhat like the one that can be seen in the second image above?

**Reply by MathiasBaumgartinger** [reply, marked as answer]
Okay, I think I finally found the unfavorable behavior about my classifier. As I am running experiments that require geo-coordinates of the underlying tiles, I used a `RandomBatchGeoSampler`. The problem with this sampler is that if the underlying datasets consists of multiple individual tiles, each batch is sampled from _one semi-randomly chosen tile only_ leading to very little diversity per batch. Even more importantly, these tiles may even be overlapping and thus semi-equal. 

The classifier thus cannot learn because the underlying distribution is so simillar that pushing negatives also pushes positives and pulling positives also pulls negatives.

I actually think `RandomBatchGeoSampler` could be much more expressive in this regard. I understand that the performance gains of this approach are good. The "random" in the name -- at least to me -- suggest something very different from what it actually does.

**Reply by adamjstewart** [reply]
Based on Figure 3A from https://arxiv.org/pdf/2111.08872, RandomBatchGeoSampler isn't even that much faster than RandomGeoSampler.

I'm in the process of rewriting all of our samplers. We might get rid of our BatchGeoSamplers unless someone has a strong need for them.

**Reply by MathiasBaumgartinger** [reply]
@adamjstewart Good to hear. Should i nonetheless open an issue to make the documentation of  `RandomBatchGeoSampler` be more verbose about this? Especially sinse `RandomGeoSampler` explicitly forwards to the `RandomBatchGeoSampler` for tile based dss?

**Reply by adamjstewart** [reply]
Sure, why not.

**Comment by adamjstewart** [comment]
P.S. Before spending too much time on #2394, you could also help review #3181, which may replace or supplement your PR.
